### PR TITLE
Assignment 04

### PR DESCRIPTION
**Pull request by Rixhee**

### CODE REVIEW

**Comment by ayushchakra** [comment]
# Rubric For Mini Project 0: JQ

## Submission

### Canvas

1: The link submitted on Canvas is for the pull request containing the
assignment submission.

0: The link submitted on Canvas is not for the pull request containing the
assignment submission.

### Git Branch

1: The branch includes only changes for this assignment.

0: The branch includes commits or changes outside this assignment.

### Git Commit Message

1: All commit messages for the branch are reasonably clear.

0: One or more commit messages for the branch are reasonably clear.

### GitHub Pull Request

1: The pull request is made to the `main` branch of the student's fork of the
course repository.

0: The pull request is made to a branch other than `main` or to a repo that is
not the student's fork of the course repository.

### Points

**Submission Score: 4/4**

## Features
### Command Line Interface
For each of the required CLI arguments `filter-string`, `path-to-json`, `color-output`, `monochrome-output`, `sort-keys`, `indent`, and `compact-output`

0.5: The argument can be correctly parsed and consumed by the rest of the package.

0: The argument is not correctly parsed and propagated to the rest of the package.

Additionally, the command line parsing correctly raises errors to the user when conflicting arguments are inputted.

1.5: Errors are properly determined and bubbled up to the user.

0: Errors are not propagated back to the user.

### Supported Filters
For each of the required filters `identity`, `object-identifier index`, `array index`, `array slice`, `pipe`, and `array iterator`

2: The filter is correctly implemented and properly parses all possible JSON inputs.

1: Minor errors exist in the implement of the filter.

0: Major errors exist in the implementation, or the filter was omitted from the package.

Additionally, comprehensive unit tests are required for each feature:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the filter, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the filter.

0: Tests are missing, or lack significant assessment of the filter's performance.

### Built-In Functions
For each of the required built-in functions `add`, `length`, `del`

2: The built-in function is correctly implemented and properly processes all possible JSON inputs.

1: Minor errors exist in the implementation of the operator.

0: Major errors exist in the implementation, or the function was omitted from the package.

Additionally, comprehensive unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the function, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the function.

0: Tests are missing, or lack significant assessment of the function's performance.

### Formatted Printing
For each of the required formats, `colored printing`, `monochrome printing`, `sorted keys`, `indents`, and `compact output`

2: The formatter is correctly implemented and properly displays any JSON object..

1: Minor errors exist in the implementation of the formatter.

0: Major errors exist in the implementation, or the formatter was omitted from the package.

Additionally, unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to ensure that the formatter can correctly handle all expected inputs.

1: Unit tests are present, but lack significant breadth of cases to fully exercise the formatter.

0: Tests are missing, or lack significant assessment of the formatter.

### Points
**Features Score: 45/61**

### Comments
1. (-1) There is a minor error in the datatype of your array iterator. If you look at the actual jq, an array iterator turns a list into a bunch of individual elements, as if they are not associated with each other. However, your iterator still format the output as if it is still in a list. Here's a concrete example of the discrepancy:
`cargo run ".[] | .name" sample_data/football.json` outputs:
```
[
  "Leo Lightning",
  "Maximus Defender",
  "Sophie Swift"
]
```
whereas, when running the same operation through jq, we get:
```
"Leo Lightning"
"Maximus Defender"
"Sophie Swift"
```
While you still correctly allow additional operation to continue element by element, this leads to bugs when using some of the built-in operators (like add).

2. (-1) When testing `test_array_slice`, you test the positive case where all the indices are within the size of the array, but don't test when they are out of bounds. This is especially important to test since you decided to error out when an out-of-bounds index is passed in.

3. (-1) Not docing points everywhere for this, but it would have been really nice if you had examined the actual contents of the error being raised through the filters to ensure that the messages are verbose and accurately describe the failure mode to the user.

4. (-1) Again, not docing points everywhere for this, but I definitely think you could have included a wider array of test cases to assess your filters, especially to ensure its robust against all datatypes.

5. (-10) No unit tests for any behavior surrounding printing.

6. (-2) Compact output should dump everything onto the same line, but print implementation still explicitly adds indents and newlines. This is a bug in `print_object` and `print_array`.

## Code Quality

### Design

#### File/Module organization

2: The program is appropriately split into reasonable files/modules

1: The program is somewhat organized, but there's room for improvement (splits aren't idiomatic, or some modules are still too large, etc.)

0: The program is not split at all

#### Scalability

5: The code is architected with scalability in mind. If more features needed to be added to the filtering/printing/CLI, they could easily be integrated.

3: The code was architected with some scalability in mind. However, significant refactoring would be needed in order to expand the feature set of this tool. (Possible symptoms include not modularizing enough/not effectively using traits + generics).

0: The code is not scalable. The entire package would need to be refactored in order to expand its features.

#### General Cleanliness

4: The code is effectively documented, especially for sources of user-facing errors. Additionally, function signatures are clear and the code is overall very readable.

2: Attempts at verbose documentation were made, but certain sources of error are unclear OR poor function signatures are used and readers would have slight difficulties while scanning the code.

0: Documentation is severely lacking in several palces and the code is not written in a readable way.

### Lint/Format

#### RustFmt

2: RustFmt produces no warnings when run on the changed files

1: RustFmt produces a few minor warnings when run on the changed files

0: RustFmt produces many warnings when run on the changed files

#### Clippy

2: Clippy produces no warnings when run on the changed files

1: Clippy produces a few minor warnings when run on the changed files

0: Clippy produces many warnings when run on the changed files

### Points
**Code Quality Score: 13/15**

Overall, I really liked how you architected your code. I'm docing 1 points for a lack of documentation throughout your project (some more function-level docstrings go a long way towards readability). I'm also taking 2 points off for scalability (see comments in the PR). 

## Summary
Add up the point totals here

**Total Score: 62/80**

File: mini-projects/00-jq-rs/src/main.rs [hunk 1]
```diff
@@ -1,3 +1,112 @@
...
+#[derive(Error, Debug)]
+enum MyErrors {
+    #[error("The indent value must be in the range of 0-7 inclusive. Current value: {0}")]
+    InvalidIndent(usize),
+
+    #[error("You can have either colored or monochrome output")]
+    ColoredAndMonochromeError,
+
+    #[error("You can have either compact or indented output")]
+    CompactAndIndentedError,
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:28]: nit: These probably could have been combined into a `ArgParsingError`, where the actual error message can be a parameter. This would make adding in new conflicting argument cases easier

File: mini-projects/00-jq-rs/src/main.rs [hunk 2]
```diff
@@ -1,3 +1,112 @@
...
+    compact_output: bool,
+
+    #[clap(required = true)]
+    needle: String,
+
+    #[clap(required = true)]
+    file: PathBuf,
+}
+
+pub fn file_path(path: PathBuf) -> Result<Value> {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:61]: nit: this function is incorrectly named

File: mini-projects/00-jq-rs/src/main.rs [hunk 3]
```diff
@@ -1,3 +1,112 @@
...
+
+    Ok(json_value)
+}
+
+fn main() -> Result<()> {
+    let args = Args::parse();
+
+    let jq_colors = env::var("JQ_COLORS").unwrap_or_else(|_| {
+        if !args.color_output || args.monochrome_output {
+            "0;0:0;0:0;0:0;0:0;0:0;0:0;0:0;0".to_string()
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:75]: Would have been nice to have this as a less opaque datatype - one example is having the actual colors and formats as n explicit hash map to what they map to

As it stands, it would be really annoying for someone to change of the types, for example if I wanted to change null to be printed red, I would have to look at the source code for how this gets parsed to know which attribute to change, rather than a well-named entry in a hash map

File: mini-projects/00-jq-rs/src/filters.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+                return Ok(FilterResult::SingleValue(sliced_value));
+            } else {
+                let indexed_value = array_index(&mutable_input, index_str)?;
+                return Ok(FilterResult::SingleValue(indexed_value));
+            }
+        } else {
+            return Err(MyErrors::ListNotFound);
+        }
+    } else if needle.starts_with(".") {
+        let value = object_identifier_filter(&mutable_input, needle)?;
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:171]: seems silly to pass in the whole needle here, rather than removing the first "." to make the `object_identifier_filter` function cleaner

File: mini-projects/00-jq-rs/src/filters.rs [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        Ok(input[index].clone())
+    } else {
+        Err(MyErrors::IndexOutOfBounds)
+    }
+}
+
+fn array_slice(input: &Value, index_str: &str) -> Result<Value, MyErrors> {
+    let indices: Vec<&str> = index_str.split(':').collect();
+    let start: usize = indices[0].parse().map_err(MyErrors::ParseError)?;
+    let end: usize = indices[1].parse().map_err(MyErrors::ParseError)?;
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:61]: what happens if someone passes in `0:1:2` to this function? Error checking on the result of the split could be helpful here

File: mini-projects/00-jq-rs/src/filters.rs [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+    if start < end && end <= input.as_array().unwrap().len() {
+        let slice = &input.as_array().unwrap()[start..end];
+        Ok(Value::Array(slice.to_vec()))
+    } else {
+        Err(MyErrors::IndexOutOfBounds)
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:67]: this doesn't seem like the correct error in the first case of the if statement - start >= end is an incorrect bounds, not out of bounds

File: mini-projects/00-jq-rs/src/filters.rs [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+    }
+}
+
+fn array_iterator(input: &Value) -> Result<Box<dyn Iterator<Item = Value>>, MyErrors> {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:71]: Why couldn't this function just return an Iterator over serde_json::Value objects?

File: mini-projects/00-jq-rs/src/filters.rs [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    if let Some(array) = input.as_array() {
+        let iter = array.clone().into_iter();
+        return Ok(Box::new(iter));
+    }
+
+    Err(MyErrors::ListNotFound)
+}
+
+pub fn pipe(input: &Value, needle: &str) -> Result<FilterResult, MyErrors> {
+    let mut current_input = input.clone(); // Clone to use it mutably
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:81]: if you wanted to make this implemented without clones, you could have just passed in a mutable reference - tradeoff here is you can't use the raw input after this function is called, but that isn't necessary in jq

File: mini-projects/00-jq-rs/src/filters.rs [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+pub enum FilterResult {
+    SingleValue(Value),
+    Iterator(Box<dyn Iterator<Item = Value>>),
+}
+
+impl PartialEq for FilterResult {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (FilterResult::SingleValue(a), FilterResult::SingleValue(b)) => a == b,
+            (FilterResult::Iterator(_), FilterResult::Iterator(_)) => true, // Just compare the variants, not the content.
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:131]: why did you choose to implement eq for iterators this way rather than checking each element?
